core_lib/src/interface/editor.cpp

```diff
@@ -445,7 +448,14 @@ void Editor::undo()
         mBackupList[mBackupIndex]->restore(this);
         mBackupIndex--;
         mScribbleArea->cancelTransformedSelection();
-        mScribbleArea->calculateSelectionRect(); // really ugly -- to improve
+
+        Layer* layer = layers()->currentLayer();
+        if (layer == nullptr) { return; }
+
+        select()->resetSelectionTransform();
+        if (layer->type() == Layer::VECTOR) {
+            setSelectionToCalculatedRect(); // still ugly... but I don't want to fiddle with the vector stuff yet
```

Comment:
Could we pass layer to `setSelectionToCalcualtedRect` so we don't have to fetch and test it again? Or better yet, get rid of that function altogether and paste the 3 lines we need from it here.

Reply:
It's not part of undo as such, so I moved it to it's own function at the time for the sake of readability but I'll move it back for now.